pybind11/_version.py

```diff
@@ -1,3 +1,35 @@
 # -*- coding: utf-8 -*-
-version_info = (2, 5, 'dev1')
-__version__ = '.'.join(map(str, version_info))
+
+import os
+
+
+DIR = os.path.abspath(os.path.dirname(__file__))
+
+
+def _to_int(s):
+    try:
+        return int(s)
+    except ValueError:
+        return s
+
+
+# Get the version from the C++ file (in-source)
+versions = {}
+
+common_h = os.path.join(
+    os.path.dirname(DIR), "include", "pybind11", "detail", "common.h"
+)
+
+with open(common_h) as f:
+    for line in f:
+        if "PYBIND11_VERSION_" in line:
+            _, name, vers = line.split()
+            versions[name[17:].lower()] = vers
+            if len(versions) >= 3:
+                break
+    else:
+        msg = "Version number not read correctly from {}: {}".format(common_h, versions)
+        raise RuntimeError(msg)
+
+__version__ = "{v[major]}.{v[minor]}.{v[patch]}".format(v=versions)
```

Comment:
I started to add bump2version, but thought single-sourcing it would be simpler. It's very important that `extras = {"global": ["pybind11==2.6.0"]}` stays in sync. If someone manually tried to increase the version and forgot that you are supposed to use a tool, you'd have `pybind11[globals]` broken in your release.

The infrastructure for including the version is the same one used for the other parts, so it doesn't add too much complexity.

Reply:
If @wjakob wants to use bumpversion always, though, I could switch. I usually don't use bumpversion, partially because adding it to the CI is an easy breaking point and overcomplicates things, and doing it manually requires you to remember to do it that way, and I don't like having to remember things.

By the way, this is closely related to the Python Packaging Authorities's "blessed" version solution, [`setuptools_scm`](https://github.com/pypa/setuptools_scm), which I'm a big fan of. It generates a `version.py` based on the git tags that gets added to the SDist, but doesn't sit in the original repo, very much like we are doing here.